services/template.go

```diff
@@ -0,0 +1,223 @@
+	return ts.sm.GetLatestTemplateByTemplateName(templateName)
+}
+
+// List lists templates.
+func (ts *templateService) List(limit int, offset int, sortBy string, order string) (state.TemplateList, error) {
+	return ts.sm.ListTemplates(limit, offset, sortBy, order)
+}
+
+// List lists templates.
+func (ts *templateService) ListLatestOnly(limit int, offset int, sortBy string, order string) (state.TemplateList, error) {
+	return ts.sm.ListTemplatesLatestOnly(limit, offset, sortBy, order)
+}
+
+// diff performs a diff between all fields (except for TemplateName and
+// Version) of two templates.
+func (ts *templateService) diff(prev state.Template, curr state.Template) bool {
+	if prev.TemplateName != curr.TemplateName {
+		return true
+	}
+	if prev.CommandTemplate != curr.CommandTemplate {
+		return true
+	}
+	if prev.Image != curr.Image {
+		return true
+	}
+	if *prev.Memory != *curr.Memory {
+		return true
+	}
+	if *prev.Gpu != *curr.Gpu {
+		return true
+	}
+	if *prev.Cpu != *curr.Cpu {
+		return true
+	}
+
+	if prev.Env != nil && curr.Env != nil {
+		prevEnv := *prev.Env
+		currEnv := *curr.Env
+		if len(prevEnv) != len(currEnv) {
+			return true
+		}
+
+		for i, e := range prevEnv {
+			if e != currEnv[i] {
+				return true
+			}
+		}
+	}
+
+	if *prev.Privileged != *curr.Privileged {
+		return true
+	}
+	if *prev.AdaptiveResourceAllocation != *curr.AdaptiveResourceAllocation {
+		return true
+	}
+	if prev.ContainerName != curr.ContainerName {
+		return true
+	}
+
+	if prev.Ports != nil && curr.Ports != nil {
+		prevPorts := *prev.Ports
+		currPorts := *curr.Ports
+		if len(prevPorts) != len(currPorts) {
+			return true
+		}
+
+		for i, e := range prevPorts {
+			if e != currPorts[i] {
+				return true
+			}
+		}
+	}
+
+	if prev.Tags != nil && curr.Tags != nil {
+		prevTags := *prev.Tags
+		currTags := *curr.Tags
+		if len(prevTags) != len(currTags) {
+			return true
+		}
+
+		for i, e := range prevTags {
+			if e != currTags[i] {
+				return true
+			}
+		}
+	}
+
+	if reflect.DeepEqual(prev.Schema, curr.Schema) == false {
+		fmt.Println("M")
+		return true
+	}
+
+	return false
+}
+
+// constructTemplateFromCreateTemplateRequest takes a CreateTemplateRequest and
+// dumps the requisite fields into a Template.
+func (ts *templateService) constructTemplateFromCreateTemplateRequest(req *state.CreateTemplateRequest) (state.Template, error) {
+	tpl := state.Template{}
+
+	if len(req.TemplateName) > 0 {
+		tpl.TemplateName = req.TemplateName
+	}
+	if req.Schema != nil {
+		tpl.Schema = req.Schema
+	}
+	if len(req.CommandTemplate) > 0 {
+		tpl.CommandTemplate = req.CommandTemplate
+	}
+	if len(req.Image) > 0 {
+		tpl.Image = req.Image
+	}
+	if req.Memory != nil {
+		tpl.Memory = req.Memory
+	}
+	if req.Gpu != nil {
+		tpl.Gpu = req.Gpu
+	}
+	if req.Cpu != nil {
+		tpl.Cpu = req.Cpu
+	}
+	if req.Env != nil {
+		tpl.Env = req.Env
+	}
+	if req.Privileged != nil {
+		tpl.Privileged = req.Privileged
+	}
+	if req.AdaptiveResourceAllocation != nil {
```

Comment:
What is the default behavior when not set?

Reply:
privileged set to false; ARA set to true